File: core/src/main/java/heast/core/network/packets/c2s/login/HelloC2SPacket.java
package heast.core.network.packets.c2s.login;

import heast.core.model.ClientInfo;
import heast.core.network.Packet;
import heast.core.network.PacketBuf;
import heast.core.network.listeners.ServerLoginListener;

public final class HelloC2SPacket implements Packet<ServerLoginListener> {

    final ClientInfo clientInfo;

    public HelloC2SPacket(ClientInfo clientInfo) {
        this.clientInfo = clientInfo;
    }

    public HelloC2SPacket(PacketBuf buf) {
        this.clientInfo = new ClientInfo(buf.readString());
    }

    @Override
    public void write(PacketBuf buf) {
        buf.writeString(clientInfo.getOs());
    }

    @Override
    public void apply(ServerLoginListener listener) {
        listener.onHello(this);
    }

    public ClientInfo getClientInfo() {
        return clientInfo;
    }
}

File: Core/network/packets/c2s/HelloC2SPacket.cs
using Core.network.listeners;

namespace Core.network.packets.c2s; 

public class HelloC2SPacket : IPacket<IServerLoginListener> {
    
    public string ClientInfo { get; }
    
    public HelloC2SPacket(string clientInfo) {
        ClientInfo = clientInfo;
    }
    
    public HelloC2SPacket(PacketBuf buf) {
        ClientInfo = buf.ReadString();
    }
    
    public void Write(PacketBuf buf) {
        buf.WriteString(ClientInfo);
    }

    public void Apply(IServerLoginListener listener) {
        listener.OnHello(this);
    }
}